### PR TITLE
chore!: make sharding configuration explicit

### PR DESCRIPTION
**Pull request by fryorcraken**
# Description

Make it explicit that the node can either be configured with _auto sharding_ or _static sharding_. Treat static sharding as an happy path (prev. was implicit error or do-nothing-happy).
Make it clear that without auto-sharding, pubsub topics can't be derived from content topics.

Do so by improving the internal node configuration and make it clear that there are two options, and that the _number of shards in the network_ is only needed for auto sharding.

Also proceed with a number of naming improvement, as per https://github.com/waku-org/specs/pull/65/

- Cluster Conf is actually a (Waku) network conf - as previously mentioned by @ivansete-status in other pr
- tcp/ip "network" conf is renamed to "endpoint conf" to avoid name clashing

Some refactoring done along the way.

# Changes

- [x] Auto sharding is enabled by default, with `1` shard in network
- [x] new `shards` (or `subscribeShards`) defaults:
  - auto sharding: all shards
  - static sharding: no shards (because we cannot guess the shard numbers)
- [x] cluster conf renamed to network conf
- [x] sharding conf is made explicit static vs auto, number of shards in network is only useful for autosharding
- [x] conf `shards` is actually the shards we want relay to subscribe too, so rename accordingly
- [x] use `uint16` for cluster id and shard id more (but not totally) consistently across the code base.
- [x] move `WebSocketConf` to the right place 

# Follow-up changes

- probably remove `activeRelayShards` in config, in favour of having `wakunode2` just automatically subscribe to the shards passed on the CLI
- Number of TODOs left in code
- define a proper type for cluster id and shard ids

## Issue

- https://github.com/waku-org/nwaku/issues/3349
- https://github.com/waku-org/nwaku/issues/3493

### CODE REVIEW

File: waku/waku_lightpush/protocol.nim [hunk 1]
```diff
@@ -26,92 +26,90 @@ type WakuLightPush* = ref object of LPProtocol
...
+      statusDesc: some(desc),
     )
-  else:
-    let pushRequest = reqDecodeRes.get()
-
-    let pubsubTopic = pushRequest.pubSubTopic.valueOr:
-      let parsedTopic = NsContentTopic.parse(pushRequest.message.contentTopic).valueOr:
-        let msg = "Invalid content-topic:" & $error
-        error "lightpush request handling error", error = msg
-        return LightpushResponse(
```
Comment (fryorcraken) [inline waku/waku_lightpush/protocol.nim]: several `return` in the error path skipped the metrics report!

File: waku/waku_lightpush/protocol.nim [hunk 2]
```diff
@@ -26,92 +26,90 @@ type WakuLightPush* = ref object of LPProtocol
...
+    return ok(handleRes.get())
+
+  return err((code: handleRes.error.code, desc: handleRes.error.desc))
 
 proc handleRequest*(
     wl: WakuLightPush, peerId: PeerId, buffer: seq[byte]
 ): Future[LightPushResponse] {.async.} =
-  let reqDecodeRes = LightpushRequest.decode(buffer)
-  var isSuccess = false
   var pushResponse: LightpushResponse
```
Comment (fryorcraken) [inline waku/waku_lightpush/protocol.nim]: Storing results in a `var` is usually a bad smell because bigger nesting is needed (instead of just exiting on an error).
Bigger nesting = less readable code.
Proof-in-pudding: some `return` were including in nesting leading to metrics being missed: https://github.com/waku-org/nwaku/pull/3468/files#r2168086052

File: waku/waku_lightpush/protocol.nim [hunk 3]
```diff
@@ -26,92 +26,90 @@ type WakuLightPush* = ref object of LPProtocol
   peerManager*: PeerManager
   pushHandler*: PushMessageHandler
   requestRateLimiter*: RequestRateLimiter
-  sharding: Sharding
+  autoSharding: Option[Sharding]
+
+proc handleRequest(
```
Comment (fryorcraken) [inline waku/waku_lightpush/protocol.nim:31]: Reviewers are free to request for this refactoring to be done in a separate PR.

File: waku/factory/internal_config.nim [hunk 4]
```diff
@@ -143,11 +137,3 @@ proc networkConfiguration*(
   )
 
   return netConfigRes
-
-# TODO: numShardsInNetwork should be mandatory with autosharding, and unneeded otherwise
```
Comment (fryorcraken) [inline waku/factory/internal_config.nim:147]: TODO Done!

File: waku/factory/waku_conf.nim [hunk 5]
```diff
@@ -166,26 +173,9 @@ proc validateNodeKey(wakuConf: WakuConf): Result[void, string] =
     return err("nodekey param is invalid")
   return ok()
 
-proc validateShards(wakuConf: WakuConf): Result[void, string] =
-  let numShardsInNetwork = wakuConf.numShardsInNetwork
-
-  # TODO: fix up this behaviour
```
Comment (fryorcraken) [inline waku/factory/waku_conf.nim:171]: Done!

File: waku/waku_core/topics/sharding.nim [hunk 6]
```diff
@@ -50,48 +51,37 @@ proc getShard*(s: Sharding, topic: ContentTopic): Result[RelayShard, string] =
 
   ok(shard)
 
-proc parseSharding*(
-    s: Sharding,
-    pubsubTopic: Option[PubsubTopic],
-    contentTopics: ContentTopic | seq[ContentTopic],
+proc getShardsFromContentTopics*(
```
Comment (fryorcraken) [inline waku/waku_core/topics/sharding.nim:54]: this function was always called when `pubsubTopic` was `none`...

File: waku/waku_lightpush/protocol.nim [hunk 7]
```diff
@@ -26,92 +26,90 @@ type WakuLightPush* = ref object of LPProtocol
...
-    if pubsubTopic.isEmptyOrWhitespace():
-      let msg = "topic must not be empty"
-      error "lightpush request handling error", error = msg
-      return LightPushResponse(
-        requestId: pushRequest.requestId,
-        statusCode: LightpushStatusCode.BAD_REQUEST.uint32,
-        statusDesc: some(msg),
-      )
+  # TODO: is that fine?
+  let res = await handleRequest(wl, peerId, pushRequest)
```
Comment (fryorcraken) [inline waku/waku_lightpush/protocol.nim]: @NagyZoltanPeter 

Is that ok to call an await here? I am not yet familiar with Nim async and whether there are any gotcha

File: tests/factory/test_external_config.nim [hunk 8]
```diff
@@ -68,11 +70,11 @@ suite "Waku config - apply preset":
 
     ## Then
     let conf = res.get()
-    check conf.shards.len == expectedConf.numShardsInNetwork.int
+    check conf.activeRelayShards.len == expectedConf.shardingConf.numShardsInCluster.int
```
Comment (SionoiS) [inline tests/factory/test_external_config.nim]: `activeRelayShards` Does that mean we can have inactive ones???

File: waku/factory/node_factory.nim [hunk 9]
```diff
@@ -137,10 +138,12 @@ proc initNode(
...
     node: WakuNode, contentTopics: seq[string]
 ): Result[seq[RelayShard], string] =
+  if node.wakuAutoSharding.isNone():
+    return err("Static sharding used, cannot get shards from content topics")
   var autoShards: seq[RelayShard]
   for contentTopic in contentTopics:
-    let shard = node.wakuSharding.getShard(contentTopic).valueOr:
-      return err("Could not parse content topic: " & error)
+    let shard = node.wakuAutoSharding.get().getShard(contentTopic).valueOr:
+        return err("Could not parse content topic: " & error)
```
Comment (SionoiS) [inline waku/factory/node_factory.nim:146]: extra tab?

**Review by SionoiS — APPROVED** [review]
LGTM

I just hope we didn't mess yet again some edge cases...

I swear this is the 5e time we refactor shard config stuff!

File: tests/factory/test_external_config.nim (hunk 8 above)
Comment (fryorcraken) [reply tests/factory/test_external_config.nim]: This is about expressing the fact that those "shards" are really shards we want to subscribe to.

Next step will be to remove that from the config entirely as it's not a node setup matter, but a run time matter.

**Comment by fryorcraken** [comment]
Started to work on fixing interops tests: https://github.com/waku-org/waku-interop-tests/pull/126

**Comment by fryorcraken** [comment]
interop-tests PR https://github.com/waku-org/waku-interop-tests/pull/126 is ready for review.

Once merged, I'll rerun tests on this PR and it should be green.

File: waku/factory/waku_conf.nim [hunk 10]
```diff
@@ -68,11 +76,10 @@ type WakuConf* {.requiresInit.} = ref object
   nodeKey*: crypto.PrivateKey
 
   clusterId*: uint16
-  shards*: seq[uint16]
+  activeRelayShards*: seq[uint16]
```
Comment (gabrielmer) [inline waku/factory/waku_conf.nim]: There's something I still don't get about `activeRelayShards`. Shouldn't the user configure which shards they want and that's it? And isn't that what it should be in the configuration, the desired shards by the user?

File: waku/waku_core/topics/sharding.nim [hunk 11]
```diff
@@ -50,48 +51,37 @@ proc getShard*(s: Sharding, topic: ContentTopic): Result[RelayShard, string] =
...
+  for content in nsContentTopics:
+    let shardsRes = s.getShard(content)
 
-        if parseRes.isErr():
-          return err("Cannot parse pubsub topic: " & $parseRes.error)
-        else:
-          parseRes.get()
+    let shard =
+      if shardsRes.isErr():
+        return err("Cannot deduce shard from content topic: " & $shardsRes.error)
```
Comment (gabrielmer) [inline waku/waku_core/topics/sharding.nim]: Maybe this was it's simpler?

```suggestion
    let shard = s.getShard(content).valueOr:
        return err("Cannot deduce shard from content topic: " & $error)
```

(and then removing the `let shardsRes = s.getShard(content)` and the ` else:
        shardsRes.get()`, which GH doesn't let me add in the suggestion)

File: waku/node/waku_node.nim [hunk 12]
```diff
@@ -322,11 +323,15 @@ proc subscribe*(
...
-
-      ($shard, some(subscription.topic))
+      if node.wakuAutoSharding.isSome():
+        let shard = node.wakuAutoSharding.get().getShard((subscription.topic)).valueOr:
+            error "Autosharding error", error = error
+            return err("Autosharding error: " & error)
+        ($shard, some(subscription.topic))
+      else:
+        let msg =
+          "Static sharding is used, relay subscriptions must specify a pubsub topic"
```
Comment (gabrielmer) [inline waku/node/waku_node.nim]: I think this msg variable is not used anywhere, same for the `ContentUnsub`case

File: waku/waku_lightpush/protocol.nim (hunk 3 above)
Comment (gabrielmer) [reply waku/waku_lightpush/protocol.nim:31]: I think it's better yes, the PR is already large enough, better to have this in a separate PR in case there's a bug in the future it will be easier to pinpoint to this

File: waku/factory/node_factory.nim [hunk 13]
```diff
@@ -290,14 +288,22 @@ proc setupProtocols(
 
     peerExchangeHandler = some(handlePeerExchange)
 
-  let autoShards = node.getAutoshards(conf.contentTopics).valueOr:
-    return err("Could not get autoshards: " & error)
+  # TODO: when using autosharding, the user should not be expected to pass any shards, but only content topics
+  # Hence, this joint logic should be removed in favour of an either logic:
+  # use passed shards (static) or deduce shards from content topics (auto)
```
Comment (gabrielmer) [inline waku/factory/node_factory.nim:293]: Not sure I understand or agree with this, one might subscribe to static shards (for example, default TWN shards) but want to use autosharding to publish messages without having to think about shards in their application's logic. Isn't that legitimate usage?

**Review by gabrielmer — COMMENTED** [review]
Woow amazing work! Thanks so much!

Added first some questions to further understand some points

File: waku/factory/waku_conf.nim (hunk 10 above)
Comment (fryorcraken) [reply waku/factory/waku_conf.nim]: spoiler: I will remove this parameter, as I am not including it in the Waku API.

these are the shards we want to subscribe to via relay when the node starts.

Define "desired shards by the user"

File: waku/factory/node_factory.nim (hunk 13 above)
Comment (fryorcraken) [reply waku/factory/node_factory.nim:293]: It's the same as https://github.com/waku-org/nwaku/pull/3468#discussion_r2176864130

I think we first need to acknowledge that these are 2 separate use cases:

(1) A developer using nwaku (as a lib or otherwise) for their app. In this case, if they use autosharding then they only want to deal with content topic to subscribe to shards or send messages

(2) A node operator running a node on TWN. Assuming said operator don't want to run a node for all shards. Why would they do that? is it to support their app? in this case, which shard should they subscribe too? 1, 2, 3? How do they know. Well, this is back to step (1). They should just deal with content topics.

File: waku/factory/waku_conf.nim (hunk 10 above)
Comment (gabrielmer) [reply waku/factory/waku_conf.nim]: > Define "desired shards by the user"

Exactly this: "the shards we want to subscribe to via relay when the node starts"

When a node is configured and the operator selects the shard they want to operate in. Why call them `activeRelayShards` instead of `shards` if it's the config desire of the user/operator and not a run time issue?

or that is the combination of both the shards specified by the user explicitly + the resolved shards by autosharding?

Comment (gabrielmer) [reply waku/factory/waku_conf.nim]: ok now I'm understanding better, that if we don't allow mixing autosharding and static sharding then makes sense!

**Review by gabrielmer — APPROVED** [review]
Thanks so much! I would wait until @Ivansete-status's review before merging just in case :)

File: waku/node/waku_node.nim [hunk 14]
```diff
@@ -112,7 +112,7 @@ type
     wakuLightpushClient*: WakuLightPushClient
     wakuPeerExchange*: WakuPeerExchange
     wakuMetadata*: WakuMetadata
-    wakuSharding*: Sharding
+    wakuAutoSharding*: Option[Sharding]
```
Comment (Ivansete-status) [inline waku/node/waku_node.nim:115]: The `wakuAutoSharding` shouldn't belong to `WakuNode`.
IMHO, this is linked to `wakuRelay` and should be there instead.

File: waku/factory/conf_builder/waku_conf_builder.nim [hunk 15]
```diff
@@ -59,8 +59,9 @@ type WakuConfBuilder* = object
   nodeKey: Option[crypto.PrivateKey]
 
   clusterId: Option[uint16]
-  numShardsInNetwork: Option[uint32]
-  shards: Option[seq[uint16]]
+  shardingConf: Option[ShardingConfKind]
+  numShardsInCluster: Option[uint16]
```
Comment (Ivansete-status) [inline waku/factory/conf_builder/waku_conf_builder.nim:63]: This is a great opportunity to have a good default.
I would not give freedom to select the number of shards in a cluster. No one is going to give a value there with certain decision.
Instead, we need to use the maximum allowed shards within a cluster, which is 1024, according to the spec: https://github.com/waku-org/specs/blob/master/standards/core/relay-sharding.md

File: waku/factory/conf_builder/waku_conf_builder.nim [hunk 16]
```diff
@@ -59,8 +59,9 @@ type WakuConfBuilder* = object
...
+  activeRelayShards: Option[seq[uint16]]
```
Comment (Ivansete-status) [inline waku/factory/conf_builder/waku_conf_builder.nim]: What are the inactive relay shards :)? I think that's been commented elsewhere: what's wrong with `shards` name?

File: waku/factory/conf_builder/waku_conf_builder.nim [hunk 17]
```diff
@@ -136,20 +137,23 @@ proc init*(T: type WakuConfBuilder): WakuConfBuilder =
...
-  b.numShardsInNetwork = some(numShardsInNetwork)
+proc withShardingConf*(b: var WakuConfBuilder, shardingConf: ShardingConfKind) =
+  b.shardingConf = some(shardingConf)
 
-proc withShards*(b: var WakuConfBuilder, shards: seq[uint16]) =
-  b.shards = some(shards)
+proc withNumShardsInCluster*(b: var WakuConfBuilder, numShardsInCluster: uint16) =
+  b.numShardsInCluster = some(numShardsInCluster)
+
+proc withActiveRelayShards*(b: var WakuConfBuilder, shards: seq[uint16]) =
```
Comment (Ivansete-status) [inline waku/factory/conf_builder/waku_conf_builder.nim]: I prefer this. More concise ;P
```suggestion
proc withShards*(b: var WakuConfBuilder, shards: seq[uint16]) =
```

File: waku/factory/conf_builder/waku_conf_builder.nim [hunk 18]
```diff
@@ -59,8 +59,9 @@ type WakuConfBuilder* = object
   nodeKey: Option[crypto.PrivateKey]
 
   clusterId: Option[uint16]
-  numShardsInNetwork: Option[uint32]
-  shards: Option[seq[uint16]]
+  shardingConf: Option[ShardingConfKind]
```
Comment (Ivansete-status) [inline waku/factory/conf_builder/waku_conf_builder.nim:62]: According to the spec, Automatic Sharding should be set by default. 
Therefore, this should not be an `Option` and be set to `AutoSharding` by default.

https://github.com/waku-org/specs/blob/master/standards/core/relay-sharding.md

File: waku/factory/conf_builder/waku_conf_builder.nim [hunk 19]
```diff
@@ -282,77 +286,89 @@ proc nodeKey(
...
+    warn "Num Shards In Cluster was provided alongside a network conf",
+      used = networkConf.shardingConf.numShardsInCluster,
+      discarded = builder.numShardsInCluster
+
+  case networkConf.shardingConf.kind
+  of StaticSharding:
+    builder.shardingConf = some(StaticSharding)
+  of AutoSharding:
+    builder.shardingConf = some(AutoSharding)
+    builder.numShardsInCluster = some(networkConf.shardingConf.numShardsInCluster)
```
Comment (Ivansete-status) [inline waku/factory/conf_builder/waku_conf_builder.nim:377]: This shouldn't be needed
```suggestion
```

File: waku/factory/conf_builder/waku_conf_builder.nim [hunk 20]
```diff
@@ -412,21 +428,36 @@ proc build*(
...
+        ShardingConf(kind: StaticSharding)
+      of AutoSharding:
+        if builder.numShardsInCluster.isSome():
+          ShardingConf(
+            kind: AutoSharding, numShardsInCluster: builder.numShardsInCluster.get()
+          )
+        else:
+          return err(
+            "Number of shards in cluster must be specified with autosharding enabled"
+          )
```
Comment (Ivansete-status) [inline waku/factory/conf_builder/waku_conf_builder.nim]: I think this can be simplified if we apply good default for `numShardsInCluster` ( `1024` ) and avoid making `shardingConf` an `Option`

File: waku/factory/external_config.nim [hunk 21]
```diff
@@ -317,10 +317,11 @@ hence would have reachability issues.""",
     # TODO: This is trying to do too much, this should only be used for autosharding, which itself should be configurable
     # If numShardsInNetwork is not set, we use the number of shards configured as numShardsInNetwork
     numShardsInNetwork* {.
-      desc: "Number of shards in the network",
+      desc:
+        "Enables autosharding and set number of shards in the cluster, set to `0` to use static sharding",
```
Comment (Ivansete-status) [inline waku/factory/external_config.nim:319]: As I mentioned before, this is a great opportunity to remove this parameter :D

File: waku/factory/networks_config.nim [hunk 22]
```diff
@@ -1,18 +1,23 @@
...
+    AutoSharding
+    StaticSharding
 
-# TODO: Rename this type to match file name
+  ShardingConf* = object
+    case kind*: ShardingConfKind
+    of AutoSharding:
+      numShardsInCluster*: uint16
+    of StaticSharding:
+      discard
```
Comment (Ivansete-status) [inline waku/factory/networks_config.nim:18]: This is beautiful indeed :heart: 
Nevertheless, I think we can simplify it by having a good default for `numShardsInCluster` (1024) in which case, the `ShardingConf` wouldn't be needed.
Sorry for repeating this same simplification idea multiple times :D

File: waku/factory/node_factory.nim (hunk 13 above)
Comment (Ivansete-status) [reply waku/factory/node_factory.nim:293]: I do agree with @gabrielmer in this point.

Autosharding seems as a technique for both distributing the traffic randomly across the available shards and to simplify the publish operation.

It seems that we will always have a combination of both static and automatic sharding. For example, _TWN_, and any other network, would not be able to operate with only automatic sharding, because the content-topic is only available when publishing a message and not when subscribing.

**Review by Ivansete-status — COMMENTED** [review]
Thanks so much for it! :raised_hands: 

Not approving yet because I didn't review it 100%

I don't want to be a blocker so feel free to merge it if you consider appropriate.
Of course, happy to revisit again in the future if you want to apply any change or split it into smaller PRs :)

Thanks again! :heart:

File: waku/node/waku_node.nim (hunk 14 above)
Comment (fryorcraken) [reply waku/node/waku_node.nim:115]: I would disagree here, light push, store and filter also uses the autosharding logic, as pubsub topic are part of the request body and needs to be inferred from the content topics.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 15 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim:63]: when you start a new network, you want to start with a small number of shards. So that you can have several users on the same shard. If you have no user on a shard, you cannot form a gossipsub network.

1024 shards by default means that you would not form any gossipsub networks, and users would not connect to each other because their application is likely to only connect to one shard.

There is a reason why we selected 8 shards for TWN.

Most new networks would be fine with 1 or 8 shards.

Also, are you thinking a default for service nodes? or for applications?

A default value of `1` may be better, for applications that don't want to use TWN.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 16 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim]: because `shards` means nothing. I guess I shall rename this variable to `relayShardsToSubscribeTo`
These are the shards we want relay to subscribe to.

Do note that this is a pure `wakunode2` service node matter. In an application, you would want the application to run `subscribe` at the start, this option is purely for CLI and I will remove it from `WakuConf`.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 17 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim]: More concise because it does not convey any meaning. Please start to see nwaku as an application library. what `shards` should a developer select? The answer is `none` when using autosharding. And for static sharding, this can be done with `subscribe` API, meaning they are the shards to subscribe to.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 18 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim:62]: I am sorry but an `Option` means that it can be set, or it can be unset.

a `default` means that if the developer does not set it, then we apply a `default value`.

Meaning there is default value, does not mean it's the only possible value.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 19 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim:377]: What do you mean it should not be needed? The number of shards in the network impact the autosharding protocol. If one node sets `8` and another sets `1024` then a given content topic `foo` will end up on different shards, meaning you won't be able to transmit messages.

File: waku/factory/external_config.nim (hunk 21 above)
Comment (fryorcraken) [reply waku/factory/external_config.nim:321]: Please do not make 3 comments about the same thing, it's a waste of time for everyone. You open the debate about removing this parameter. We can discuss it in one comment thread.

File: waku/factory/node_factory.nim (hunk 13 above)
Comment (fryorcraken) [reply waku/factory/node_factory.nim:293]: A given network should either have static sharding, or auto sharding, not both. Do we agree with that?

I am a developer that is building my new `foobar` app on TWN, hence auto sharding is enabled.

Please walk me through how I need to select the "right shard" to subscribe to.

File: waku/factory/waku_conf.nim (hunk 10 above)
Comment (fryorcraken) [reply waku/factory/waku_conf.nim]: > When a node is configured and the operator selects the shard they want to operate in. Why call them `activeRelayShards` instead of `shards` if it's the config desire of the user/operator and not a run time issue?

Because the action is unclear, why does one need to specify `shards` ? Well, because that the shards we want relay to subscribe to. But that only applies in one specific case: when running a service node with relay.

When running an application, the users or developer does **not** subscribe to shards, it subscribe to specific content topics they want to receive messages on.

And when not using relay, they don't need shards either.

And even in this case, there is still the problem of:

- I am using TWN
- I want to run a service node to support my node
- What "shard" do I select?

Well, you shouldn't need to "select" a shard, but a "content topic" instead (or a partial one), because autosharding abstract the concept of shards.

File: waku/waku_lightpush/protocol.nim (hunk 3 above)
Comment (fryorcraken) [reply waku/waku_lightpush/protocol.nim:31]: https://github.com/waku-org/nwaku/pull/3486

File: waku/factory/waku_conf.nim (hunk 10 above)
Comment (fryorcraken) [reply waku/factory/waku_conf.nim]: > not a run time issue?

So we are clear, it is a run time issue. Getting relay to subscribe to shards is not an initialization matter, it is a run time matter.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 18 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim:62]: Also well spotted that there is a gap in the spec IMO. It says that auto sharding is the default choice, but does not recommend autosharding parameters such as the number of shards.

Cc @jm-clius: what should be the default number of shards, in the same vein as

> Autosharding selects shards automatically and is the default behavior for shard choice

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 19 above)
Comment (Ivansete-status) [reply waku/factory/conf_builder/waku_conf_builder.nim:377]: Yes indeed, you are absolutely right. When I wrote my comment I was thinking about having 1024 by default to all nodes but, subscribing to 1024 shards doesn't seem like a great idea because I suspect it won't perform very well.
The static and auto shards generates a lot of confusion and is a source of possible discrepancies among nodes. For example, what happens if someone configures the following?

- shards [1, 2, 3]
  plus
- numShardsInCluster=4

In that case, the `foo` content-topic can potentially fall in wrong shard (considering all nodes are in 1, 2, 3.)

Therefore, I think we need to infer the value of `numShardsInCluster`, or have a good default.

Another possible option would be to implement a hybrid solution between static and auto sharding: always have defined the `shards` list and, in case of using auto-sharding, infer the available shards from that available `shards` list. wdyt?

Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim:377]: Let's take it from a user PoV. I am only covering when `auto-sharding` is enabled.

1. Library user/application developer: in this case, I don't think `--shards` should be used at all. They should use `subscribe (contentTopic)` to subscribe to a topic.
When sending a message,  then we will have two choices when they are not subscribed to the right shard:
  1. non-blocking logic: fallback to light push
  2. blocking logic: error and require `subscribe` to be done first
2. Node operator for a given application: In this case, the operator want to support their specific application and run relay/store/etc for their shard. which means they would use `--content-topics` to subscribe to the right shards
3. Node operators - for Waku/incentivization/etc: In this instance, a node operator may want to support all shards, or specific shards where there is most traffic. Meaning `--shards` is most appropriate.

## In conclusion

To setup relay shard subscription:

A. On the `wakunode2` CLI, it is fine for a user to either both `--shards` or `--content-topic` to specify what shards to subscribe to, but it should not be both *at the same time*. Note that a _partial content topic_ should be accepted (only first fields are needed for autosharding).
B. On the `Waku API`, then I would suggest that neither `--shards` nor `--content-topic` are available at node initialization, but instead, `subscribe(contentTopic)` is used.

When sending a message, in relay mode, on a shard not subscribed to. I suggest to `error out` for now. We can decide at a later stage to fallback to light push.

Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim:377]: > Therefore, I think we need to infer the value of numShardsInCluster, or have a good default.

The default nework conf value should be "TWN", so that this number is 8.
If someone want to roll out their own network, and own cluster id, then I agree we can propose a default value. Probably 1 or 4. 
@jm-clius WDYT?

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 20 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim]: `ShardingConf` MUST remain an enum. Because `numShardsInCluster` does not make sense for a static sharding conf.

File: waku/waku_lightpush/protocol.nim (hunk 3 above)
Comment (fryorcraken) [reply waku/waku_lightpush/protocol.nim:31]: rebased

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 19 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim:377]: @Ivansete-status just so we are on the same page, you seem to say that `numShardsInCluster` is not needed, as well as wanting a good default, in lieu of the value.

Ultimately, it **is** needed because different network may have different shards. For example, we may grown the number of shards on Status network to 50 with the chat sdk. But someone bootstrapping a brand new network may prefer to stick to `1` until scaling is needed. So this parameter **is** needed.
Then the question of default value remain up to discussion, right?

Comment (Ivansete-status) [reply waku/factory/conf_builder/waku_conf_builder.nim:377]: A. and B. makes sense to me in this comment: https://github.com/waku-org/nwaku/pull/3468#discussion_r2181398151
The following looks great to me as well:
> When sending a message, in relay mode, on a shard not subscribed to. I suggest to error out for now.

------

> If someone want to roll out their own network, and own cluster id, then I agree we can propose a default value. Probably 1 or 4.

Why not 8 too?

I think we need a good first default, 8, and then, define an scaling algorithm that makes the network breath, i.e., subscribe to more or less shards depending on bandwidth needed. That will need a consensus in the network.

For that, I believe we need to also establish a fixed _consensus_ shard on every network, on every cluster. For example, we can enforce that every cluster will have the shard `0` to be reserved for only containing Waku consensus and coordination messages. (Another option is having `waku/2/rs/consensus` pubsubtopic and leave shard `0` for data.)

That consensus topic will help nodes to reach an agreement in:
- Define the total number of shards in the network. Each node will subscribe/unsubscribe accordingly.
- Determine which nodes will act as bridge or aggregator between shards.
- When to start scaling up/down, i.e., when to start changing the current shard set.
- Define the shards a certain node will subscribe to (we can do math considering the peerId.)

------

> Ultimately, it is needed because different network may have different shards.

I think you are absolutely right with that but, on the other hand, I believe we need to implement a network algorithm that makes the network scale up/down accordingly. I strongly suggest following the `consensus` topic idea that I commented above.

> Then the question of default value remain up to discussion, right?

Yes :) I'd suggest having a good default plus an algorithm that makes it adaptive as per network conditions.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 18 above)
Comment (jm-clius) [reply waku/factory/conf_builder/waku_conf_builder.nim:62]: This is because Autosharding is only defined as a concept in this spec and presented as the preferred default option when considering sharding for Waku networks, followed by static (and even named sharding, which we don't even implement anymore). This spec does not define any specific application of autosharding for specific Waku Network(s), as this more properly belongs to the "consensus" specification for a specific network (e.g. the Status Network(s), TWN, etc.). The _only_ network for which we have defined autosharding is TWN, as it's the only specified network with a consensus on cluster ID and number of shards: https://rfc.vac.dev/waku/standards/core/64/network#autosharding.
This out-of-band consensus on a network definition is a **must** for autosharding to work as expected. The spec should indeed be clarified in this regard.

My initial preference was that autosharding, as an app-level API simplification, should only be available when TWN is configured. However, it became clear for testing purposes and more modular use of Waku, that people preferred having the option of ad-hoc creating their own "network definitions" with autosharding by configuring a `numShardsInCluster` option. Note that this is a higher-level **network definition** and has nothing to do at all with which relay shards are actually being subscribed to. You could perfectly well set this parameter and not subscribe to any traffic at all. Also, it implies that these experimenters "know" that others participating in the experiment have configured their nodes in the same way (or are using the default config - see below). The way to formalise this is in the form of published network specifications for each new network, where the number of shards, services, etc. are agreed upon.

Autosharding also only makes sense in the API when applications build on top of a Waku node. In other words, it should only be used when the node is run within specific content topic (i.e. "App") context - it has no use for general service nodes that do not care about specific content topics.

My view of defaults:
Default configuration should always be for TWN.
If someone sets a different cluster ID, the default should be that that cluster is assumed to have only `1` shard. There's no need to pre-emptively "scale" these clusters - we had a specific reason for making TWN experimentally work with 8 out of the box, but that reasoning doesn't apply to the defaults.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 19 above)
Comment (jm-clius) [reply waku/factory/conf_builder/waku_conf_builder.nim:377]: I don't think at this stage we need a consensus algorithm to set a network definition. Consensus for the networks we define is reached by publishing an RFC. We have only one defined network, which is TWN for which number of shards is 8. If we need to scale, we'll publish a new version of the spec - TWN Gen 1 if you will - and autosharding is already defined in such a way that it will work with backwards compatibility.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 18 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim:62]: Thanks @jm-clius. I actually came to the same conclusion overnight. Which hopefully is in line with @Ivansete-status 's original comment.

With the introduction of the `preset` concept, we are moving towards TWN being the default. It is not straightforward and as RLN plays a part in it. 

I suggest to have nwaku default to autosharding with `1` shard when TWN is not set. For a few reason:

1. Autosharding means abstracting shards concept, which is better for dev ex
2. Autosharding is yet to be properly dogfood on large applications, TWN with `8` is a starter but no serious app is running on it. Hence, a value of `1` is low risk.

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 19 above)
Comment (fryorcraken) [reply waku/factory/conf_builder/waku_conf_builder.nim:377]: As per https://github.com/waku-org/nwaku/pull/3468#discussion_r2182728357 I will proceed with the following changes:

1. autosharding by default
2. num of shards in cluster: `1`

File: waku/factory/conf_builder/waku_conf_builder.nim (hunk 18 above)
Comment (Ivansete-status) [reply waku/factory/conf_builder/waku_conf_builder.nim:62]: Thanks for the feedback! I with the suggested approach :ok_hand:
